docs/_docs/02_testing-best-practices/unit-integration-end-to-end-test.md

```diff
@@ -22,3 +22,178 @@ Link to the video at [infoq.com](https://www.infoq.com/presentations/automated-t
+
+**_TODO_**: As this is a one day project I will only attempt to do step 3, 4, and 6 today. Time permitting I will add 2
+as well 
+
+1. Copy your terraform code to a temporary folder
+2. Optionally call a setup function you provide
+3. Init and Apply the terraform code
+4. Call a list of validators you provide
+5. Optionally call a tearDown function you provide
+6. Destroy the terraform code
+
+Most of these steps are wrapped into stages which you can skip (see [Iterating locally using test stages]({% link _docs/02_testing-best-practices/iterating-locally-using-test-stages.md %})). This means you can skip individual stages
+by setting env var SKIP_\<stage_name\> to any value.
+
+If you skip the init_apply step then UnitTest will automatically load your previously applied outputs from the working directory.
+
+It's common to need to pass data around between validators (for example the awsRegion, or random names you generated to
+namespace that test run), so the validation function also receives an arg testData (which is defined as the [empty interface](https://tour.golang.org/methods/14)). You can define your own struct to pass data around
+and cast this argument back to your struct to ensure type safety and give you access to shared data between tests.
+
+The full list of stages is:
+
+* setup
+* init_apply
+* validate
+* teardown
+* destroy
+
+Signature of UnitTest:
+
+```go
+// The output of the module after it has been applied, as returned by terraform.OutputAll
+type ModuleOutputs map[string]interface{}
+
+// A validation function to test your terraform code
+type Validator func(t *testing.T, moduleOutputs terraform.ModuleOutputs, testData interface{}, workingDir string)
+
+type UnitTestConfig struct {
+  TerraformOptions *terraform.Options
+  // If you do not provide a working directory it will default to ./.terratest-unit-test/TEST_FUNCTION_NAME/
+  WorkingDirectory string
+  Setup func(t *testing.T, workingDir string)
+  TearDown func(t *testing.T, moduleOutputs terraform.ModuleOutputs, workingDir string)
+  Validators []Validator
+  TestData interface{}
+}
+
+func UnitTest(t *testing.T, testConfig *terraform.UnitTestConfig)
+```
+
+Examples:
+
+This example shows the simplest example with multiple Validators, with the default working directory
+```go
+func TestWebServer(t *testing.T) {
+  terraformOptions := &terraform.Options {
+    // The path to where your Terraform code is located
+    TerraformDir: "../web-server",
+  }
+
+  terrform.UnitTest(&terraform.UnitTestConfig{
+    TerraformOptions: terraformOptions,
+    Validators: []terraform.Validator{
+      // A list of validation functions to test your module
+      validateWebServer,
+      validateImageHosting,
+      validateSSLApplied,
+    },
+  })
+}
+
+func validateWebServer(t *testing.T, outputs terraform.ModuleOutputs, testDataStruct interface{}, workingDir string) {
+}

```

Comment:
NIT: Should have a comment indicating that `validateImageHosting` and `validateSSLApplied` are additional functions defined here that will have the same signature.